scripts/reduce_dimensions.R

```
@@ -36,77 +47,87 @@ METHOD <- snakemake@wildcards$dim_red
 # Sparse PLS
 #
 if (METHOD == 'sparse_pls') {
-    library(spls)
-    set.seed(2897)
-    # columns with too many zeros can end up having zero variance in the inner CV, which makes SPLS unhappy
-    zero_fvars <- apply(feat[train_idx,], 2, function(col) sum(col != 0))/sum(train_idx) <= 0.15
-    if (sum(zero_fvars) > 0) {
-        message(paste0("Excluding ", sum(zero_fvars), " features with large number of 0s."))
-	feat <- feat[, !zero_fvars]
-    }
-    # Select number of components and sparsity penalization (eta) by cross-validation
-    cv_chs <- cv.spls(feat[train_idx, ], y[train_idx],
-                               eta = c(seq(0.3, 0.7, by = 0.1), 0.75),  K = c(1:10), fold = 10)
-
-    # Fit sparse PLS using selected number of components and eta
-    spfit <- spls(feat[train_idx, ], y[train_idx], eta = cv_chs$eta.opt, K = cv_chs$K.opt)
-    
-    # Project entire feature matrix onto principal components
-    selected <- rownames(spfit$projection)
-    latent_feats <- feat[,selected] %*% spfit$projection
-    colnames(latent_feats) <- paste0("LF", 1:ncol(latent_feats))
-
-    # save projection matrix
-    projection <- t(spfit$projection)
-    colnames(projection) <- selected 
-    write_tsv(as.data.frame(projection), snakemake@output[[2]])
+  library(spls)
+  set.seed(2897)
+
+
+  # columns with too many zeros can end up having zero variance in the inner CV, which makes SPLS unhappy
+  # define lower bound for number of non-zero entries
+  MIN_NONZERO_PERC <- 0.2
```

Should this be made into a config parameter so that is can be changed without modifying the code?

---

Yeah, that would be good. I actually want to move away from using this package purely because of this issue--it's quite frustrating when I don't see any reason why they should throw an error for having a feature with 0 variance in the dataset. And because it's based on random splits of the data it's hard to really prevent the error being thrown across all of CV iterations.